docs/examples/cart-join.md

```diff
@@ -0,0 +1,256 @@
+attached, so in this case we have `(user_id, joined_state)`. Let's
+format that into a dictionary for output.
+
+```python
+def format_output(user_id__joined_state):
+    user_id, joined_state = user_id__joined_state
+    return {
+        "user_id": user_id,
+        "paid_order_ids": joined_state["paid_order_ids"],
+        "unpaid_order_ids": joined_state["unpaid_order_ids"],
+    }
+
+flow.map(format_output)
+```
+
+And finally, capture this output and send it to our output builder.
+
+```python
+flow.capture()
+```
+
+On that note, we'll serialize the final output as JSON and print it.
+
+```python
+def output_builder(worker_index, worker_count):
+    def output_handler(epoch_item):
+        epoch, item = epoch_item
+        line = json.dumps(item)
+        print(epoch, line)
+    return output_handler
+```
+
+## Execution
+
+Now let's run our dataflow without state recovery:
+
+```python doctest:IGNORE_EXCEPTION_DETAIL doctest:ELLIPSIS
+from bytewax import run_main
+
+run_main(flow, input_builder, output_builder)
+```
+
+Cool, we're chugging along and then OH NO!
+
+```{testoutput}
+0 {"user_id": "a", "paid_order_ids": [], "unpaid_order_ids": [1]}
+1 {"user_id": "a", "paid_order_ids": [], "unpaid_order_ids": [1, 2]}
+2 {"user_id": "b", "paid_order_ids": [], "unpaid_order_ids": [3]}
+Traceback (most recent call last):
+...
+json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
+```
+
+Something went wrong! In this case it was that we had a non-JSON line
+`FAIL HERE` in the input, but you could imagine that Kafka consumer
+breaks or the VM is killed or something else bad happens!
+
+We've also built up very valuble state in our stateful map operator
+and we don't want to pay the penalty of having to re-read our input
+all the way from the beginning to build it back up. Let's see how we
+could have implemented state recovery to allow that to happen in the
+future!
+
+## Recovery Prep
+
+Following our checklist in [Bytewax's documentation on recovery]() we
+need to enhance our input builder with a few things.
+
+First, we need the ability to resume from an arbitrary epoch. Since
+we're using the line in the file as the epoch, we can skip ahead to
+that line.
+
+```python
+RESUME_FROM_EPOCH = 0
+
+
+def input_builder(worker_index, worker_count):
+    assert worker_index == 0  # We're not going to worry about multiple workers yet.
+    with open("cart-join.jsons") as f:
+        for epoch, line in enumerate(f):
+            if epoch < RESUME_FROM_EPOCH:
+                continue
+            obj = json.loads(line)
+            yield epoch, obj
+```
+
+
+We need a **recovery store**. For this example we'll use a local
+[Sqlite]() database because we're running on a single machine.
+
+
+```python
+from tempfile import TemporaryDirectory
+from bytewax.recovery import SqliteRecoveryConfig
+
+recovery_dir = TemporaryDirectory()  # We'll store this somewhere temporary for this test.
```

Comment:
Hmmm. Perhaps we should provide a more formal way of saying "I know my recovery state is empty; create it anyway" and error out otherwise? This is a confusing manifestation of that.